Makefile

```diff
@@ -1,85 +1,100 @@
+COVERAGE?=$(CURDIR)/coverage.html
+SC_PKG=github.com/kubernetes-incubator/service-catalog
+TOP_DIRS=cmd contrib pkg
+SRC_DIRS=$(shell sh -c "find $(TOP_DIRS) -name \\*.go -exec dirname {} \\; | sort | uniq")
+TEST_DIRS=$(shell sh -c "find $(TOP_DIRS) -name \\*_test.go -exec dirname {} \\; | sort | uniq")
+SRC_PKGS=$(addprefix $(SC_PKG)/,$(SRC_DIRS))
+
+GO_VERSION=1.7.3
+GO_BUILD=go build
+
+# This section builds the output binaries
+#########################################
+build: init \
+       $(BINDIR)/controller $(BINDIR)/registry $(BINDIR)/k8s-broker \
+       $(BINDIR)/service-catalog $(BINDIR)/user-broker \
+       $(BINDIR)/apiserver
+
+controller: $(BINDIR)/controller
+$(BINDIR)/controller: pkg/controller/catalog
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
+
+registry: $(BINDIR)/registry
+$(BINDIR)/registry: contrib/registry
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
 
-# Use this target when you want to build everything using docker containers.
-# Good for cases when you don't have the tools installed (like glide or go).
-docker-all: .scBuildImage
-	docker run --rm -ti \
-	  -v $(PWD):/go/src/github.com/kubernetes-incubator/service-catalog \
-	  scbuildimage \
-	  make .dockerInit all
-
-# .dockerInit tells us if our vendor stuff if out of date or not.
-# And if so we'll run init under our docker build.  For non-docker builds
-# it is assumed you'll run "make init" manually.
-.dockerInit: glide.yaml
-	make init
-	echo > .dockerInit
-
-# .scBuildImage is used to know when the docker image ("scbuildimage") is out
-# of date with the Dockerfile.
+$(BINDIR)/k8s-broker: contrib/broker/k8s
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
+
+$(BINDIR)/service-catalog: cmd/service-catalog
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
+
+$(BINDIR)/user-broker: contrib/broker/k8s
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
+
+apiserver: $(BINDIR)/apiserver
+$(BINDIR)/apiserver: cmd/service-catalog
+	$(DOCKER) $(GO_BUILD) -o $@ $(SC_PKG)/$^
+
+# Some prereq stuff
+###################
+init: .scBuildImage .init
+
+# .init is used to know when some Glide dependencies are out of date
+.init: glide.yaml
+	$(DOCKER) glide install --strip-vendor
+	echo > $@
+
+# .scBuildImage tells when the docker image ("scbuildimage") is out of date
 .scBuildImage: hack/Dockerfile
 	docker build -t scbuildimage - < hack/Dockerfile
-	echo > .scBuildImage
+	echo > $@
+
+# Util targets 
+##############
 
-# Build the same target recursively in all directories.
-$(SUB): %.sub:
-	$(ECHO) for dir in $(DIRS); do $(MAKE) --no-print-directory -C "$${dir}" $* || exit $$? ; done
+verify: init
+	@echo Running verify:
+	@for i in $(TOP_DIRS); do \
+      $(DOCKER) gofmt -l $$i || exit $$? ; \
+	done > .out 2>&1
+	@bash -c '[ "`cat .out`" == "" ] || \
+	  (echo -e "\n*** Please 'gofmt' the following:" ; cat .out ; echo ; false)'
+	@rm .out
+	@for i in $(SRC_PKGS); do \
+	  ($(DOCKER) golint --set_exit_status $$i/... && \
+	  $(DOCKER) go vet $$i )|| exit $$? ; \
```

Comment:
I don't suppose there's any black `Makefile` magic to run these (and other applicable) steps in parallel?